File: examples/go/core/shared/app_host_contracts.go
package shared

type DomainFactory func() any
type ContextualFactory func(ctx any) any

type AppCaseSurfaces struct {
	Domain  DomainFactory
	API     ContextualFactory
	UI      ContextualFactory
	Stream  ContextualFactory
	Agentic ContextualFactory
}

type AgenticCaseRef struct {
	Domain   string `json:"domain"`
	CaseName string `json:"caseName"`
}

type AgenticCatalogEntry struct {
	Ref                  AgenticCaseRef `json:"ref"`
	PublishedName        string         `json:"publishedName"`
	Definition           any            `json:"definition"`
	IsMcpEnabled         bool           `json:"isMcpEnabled"`
	RequiresConfirmation bool           `json:"requiresConfirmation"`
	ExecutionMode        string         `json:"executionMode"`
}

type AppRegistry interface {
	Cases() map[string]map[string]AppCaseSurfaces
	Providers() Dict
	Packages() Dict
}

type AgenticRegistry interface {
	AppRegistry
	ListAgenticCases() []AgenticCaseRef
	GetAgenticSurface(ref AgenticCaseRef) ContextualFactory
	InstantiateAgentic(ref AgenticCaseRef, ctx any) (any, error)
	BuildCatalog(ctx any) ([]AgenticCatalogEntry, error)
	ResolveTool(toolName string, ctx any) (*AgenticCatalogEntry, error)
	ListMcpEnabledTools(ctx any) ([]AgenticCatalogEntry, error)
}

File: examples/java/core/shared/AppHostContracts.java
package core.shared;

import core.BaseAgenticCase;
import core.BaseAgenticCase.AgenticContext;
import core.BaseAgenticCase.AgenticDefinition;
import core.BaseApiCase.ApiContext;
import core.BaseStreamCase.StreamContext;
import core.BaseUiCase.UiContext;
import java.util.List;
import java.util.Map;

public final class AppHostContracts {
  private AppHostContracts() {}

  @FunctionalInterface
  public interface DomainFactory {
    Object create();
  }

  @FunctionalInterface
  public interface ApiFactory {
    Object create(ApiContext context);
  }

  @FunctionalInterface
  public interface UiFactory {
    Object create(UiContext context);
  }

  @FunctionalInterface
  public interface StreamFactory {
    Object create(StreamContext context);
  }

  @FunctionalInterface
  public interface AgenticFactory {
    Object create(AgenticContext context);
  }

  public static class AppCaseSurfaces {
    public DomainFactory domain;
    public ApiFactory api;
    public UiFactory ui;
    public StreamFactory stream;
    public AgenticFactory agentic;

    public AppCaseSurfaces domain(DomainFactory value) {
      this.domain = value;
      return this;
    }

    public AppCaseSurfaces api(ApiFactory value) {
      this.api = value;
      return this;
    }

    public AppCaseSurfaces ui(UiFactory value) {
      this.ui = value;
      return this;
    }

    public AppCaseSurfaces stream(StreamFactory value) {
      this.stream = value;
      return this;
    }

    public AppCaseSurfaces agentic(AgenticFactory value) {
      this.agentic = value;
      return this;
    }
  }

  public interface AppRegistry {
    Map<String, Map<String, AppCaseSurfaces>> cases();
    Map<String, Object> providers();
    Map<String, Object> packages();
  }

  public record AgenticCaseRef(String domain, String caseName) {}

  public static class AgenticCatalogEntry {
    public AgenticCaseRef ref;
    public String publishedName;
    public AgenticDefinition<?, ?> definition;
    public boolean isMcpEnabled;
    public boolean requiresConfirmation;
    public String executionMode;
  }

  public interface AgenticRegistry extends AppRegistry {
    List<AgenticCaseRef> listAgenticCases();
    AgenticFactory getAgenticSurface(AgenticCaseRef ref);
    BaseAgenticCase<?, ?> instantiateAgentic(AgenticCaseRef ref, AgenticContext ctx);
    List<AgenticCatalogEntry> buildCatalog(AgenticContext ctx);
    AgenticCatalogEntry resolveTool(String toolName, AgenticContext ctx);
    List<AgenticCatalogEntry> listMcpEnabledTools(AgenticContext ctx);
  }
}
